File: Sources/Utils/ArchiveExtractor.swift
import Foundation
import ZIPFoundation

/// Utility class for extracting archive files (APK, AAR, AAB, JAR)
/// All these formats are ZIP-based archives
class ArchiveExtractor {

    // MARK: - Types

    struct ExtractionResult {
        let success: Bool
        let extractedPath: String?
        let filesExtracted: Int
        let errorMessage: String?

        static func success(path: String, fileCount: Int) -> ExtractionResult {
            ExtractionResult(success: true, extractedPath: path, filesExtracted: fileCount, errorMessage: nil)
        }

        static func failure(message: String) -> ExtractionResult {
            ExtractionResult(success: false, extractedPath: nil, filesExtracted: 0, errorMessage: message)
        }
    }

    struct ArchiveInfo {
        let fileName: String
        let fileSize: Int64
        let entryCount: Int
        let previewEntries: [String]
    }

    // MARK: - Public Methods

    /// Extract an archive file to a destination directory
    /// - Parameters:
    ///   - archiveURL: The archive file to extract (APK, AAR, AAB, or JAR)
    ///   - destinationURL: The directory where files will be extracted
    ///   - onProgress: Optional callback for progress updates
    /// - Returns: ExtractionResult containing the result information
    func extractArchive(
        from archiveURL: URL,
        to destinationURL: URL,
        onProgress: ((String) -> Void)? = nil
    ) -> ExtractionResult {
        // Validate input
        guard FileManager.default.fileExists(atPath: archiveURL.path) else {
            return .failure(message: "Archive file does not exist: \(archiveURL.path)")
        }

        guard isValidArchive(archiveURL) else {
            return .failure(message: "Not a valid archive file. Supported formats: APK, AAR, AAB, JAR")
        }

        // Create destination directory if it doesn't exist
        do {
            try FileManager.default.createDirectory(
                at: destinationURL,
                withIntermediateDirectories: true,
                attributes: nil
            )
        } catch {
            return .failure(message: "Failed to create destination directory: \(error.localizedDescription)")
        }

        // Extract the ZIP archive
        var filesExtracted = 0

        do {
            guard let archive = Archive(url: archiveURL, accessMode: .read) else {
                return .failure(message: "Failed to open archive file")
            }

            for entry in archive {
                let entryName = entry.path
                onProgress?("Extracting: \(entryName)")

                let destURL = destinationURL.appendingPathComponent(entryName)

                // Prevent zip-slip vulnerability
                guard destURL.path.hasPrefix(destinationURL.path) else {
                    return .failure(message: "Potential zip-slip attack detected in entry: \(entryName)")
                }

                if entry.type == .directory {
                    try FileManager.default.createDirectory(
                        at: destURL,
                        withIntermediateDirectories: true,
                        attributes: nil
                    )
                } else {
                    // Create parent directories if needed
                    let parentURL = destURL.deletingLastPathComponent()
                    try FileManager.default.createDirectory(
                        at: parentURL,
                        withIntermediateDirectories: true,
                        attributes: nil
                    )

                    // Extract file
                    _ = try archive.extract(entry, to: destURL)
                    filesExtracted += 1
                }
            }

            onProgress?("Extraction complete!")

            return .success(path: destinationURL.path, fileCount: filesExtracted)

        } catch {
            return .failure(message: "Error extracting archive: \(error.localizedDescription)")
        }
    }

    /// Get information about an archive without extracting it
    /// - Parameter archiveURL: The archive file to inspect
    /// - Returns: ArchiveInfo or nil if the file cannot be read
    func getArchiveInfo(from archiveURL: URL) -> ArchiveInfo? {
        guard FileManager.default.fileExists(atPath: archiveURL.path),
              isValidArchive(archiveURL) else {
            return nil
        }

        do {
            guard let archive = Archive(url: archiveURL, accessMode: .read) else {
                return nil
            }

            var entryCount = 0
            var entries: [String] = []

            for entry in archive {
                entryCount += 1
                if entries.count < 100 { // Limit preview to first 100 entries
                    entries.append(entry.path)
                }
            }

            let attributes = try FileManager.default.attributesOfItem(atPath: archiveURL.path)
            let fileSize = attributes[.size] as? Int64 ?? 0

            return ArchiveInfo(
                fileName: archiveURL.lastPathComponent,
                fileSize: fileSize,
                entryCount: entryCount,
                previewEntries: entries
            )

        } catch {
            return nil
        }
    }

    // MARK: - Private Methods

    /// Check if the file is a valid archive (APK, AAR, AAB, or JAR)
    private func isValidArchive(_ url: URL) -> Bool {
        let validExtensions = ["apk", "aar", "aab", "jar"]
        return validExtensions.contains(url.pathExtension.lowercased())
    }
}

File: src/main/kotlin/utils/ArchiveExtractor.kt
package utils

import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.util.zip.ZipEntry
import java.util.zip.ZipInputStream

/**
 * Utility class for extracting archive files (APK, AAR, AAB, JAR)
 * All these formats are ZIP-based archives
 */
class ArchiveExtractor {

    data class ExtractionResult(
        val success: Boolean,
        val extractedPath: String? = null,
        val filesExtracted: Int = 0,
        val errorMessage: String? = null
    )

    /**
     * Extract an archive file to a destination directory
     * @param archiveFile The archive file to extract (APK, AAR, AAB, or JAR)
     * @param destinationDir The directory where files will be extracted
     * @param onProgress Optional callback for progress updates
     * @return ExtractionResult containing the result information
     */
    fun extractArchive(
        archiveFile: File,
        destinationDir: File,
        onProgress: ((String) -> Unit)? = null
    ): ExtractionResult {
        try {
            // Validate input
            if (!archiveFile.exists()) {
                return ExtractionResult(
                    success = false,
                    errorMessage = "Archive file does not exist: ${archiveFile.absolutePath}"
                )
            }

            // Check if the path is a directory instead of a file
            if (archiveFile.isDirectory) {
                return ExtractionResult(
                    success = false,
                    errorMessage = "Selected path is a directory, not a file. Please select an archive file (APK, AAR, AAB, or JAR)."
                )
            }

            if (!isValidArchive(archiveFile)) {
                return ExtractionResult(
                    success = false,
                    errorMessage = "Not a valid archive file. Supported formats: APK, AAR, AAB, JAR"
                )
            }

            // Create destination directory if it doesn't exist
            if (!destinationDir.exists()) {
                destinationDir.mkdirs()
            }

            var filesExtracted = 0

            // Extract the ZIP archive
            ZipInputStream(FileInputStream(archiveFile)).use { zipInputStream ->
                var entry: ZipEntry? = zipInputStream.nextEntry

                while (entry != null) {
                    val entryName = entry.name
                    onProgress?.invoke("Extracting: $entryName")

                    val destFile = File(destinationDir, entryName)

                    // Prevent zip-slip vulnerability
                    if (!destFile.canonicalPath.startsWith(destinationDir.canonicalPath)) {
                        return ExtractionResult(
                            success = false,
                            errorMessage = "Potential zip-slip attack detected in entry: $entryName"
                        )
                    }

                    if (entry.isDirectory) {
                        destFile.mkdirs()
                    } else {
                        // Create parent directories if needed
                        destFile.parentFile?.mkdirs()

                        // Extract file
                        FileOutputStream(destFile).use { outputStream ->
                            zipInputStream.copyTo(outputStream)
                        }
                        filesExtracted++
                    }

                    zipInputStream.closeEntry()
                    entry = zipInputStream.nextEntry
                }
            }

            onProgress?.invoke("Extraction complete!")

            return ExtractionResult(
                success = true,
                extractedPath = destinationDir.absolutePath,
                filesExtracted = filesExtracted
            )

        } catch (e: Exception) {
            return ExtractionResult(
                success = false,
                errorMessage = "Error extracting archive: ${e.message}"
            )
        }
    }

    /**
     * Check if the file is a valid archive (APK, AAR, AAB, or JAR)
     */
    private fun isValidArchive(file: File): Boolean {
        val extension = file.extension.lowercase()
        return extension in listOf("apk", "aar", "aab", "jar")
    }

    /**
     * Get information about an archive without extracting it
     */
    fun getArchiveInfo(archiveFile: File): ArchiveInfo? {
        try {
            if (!archiveFile.exists() || archiveFile.isDirectory || !isValidArchive(archiveFile)) {
                return null
            }

            var entryCount = 0
            val entries = mutableListOf<String>()

            ZipInputStream(FileInputStream(archiveFile)).use { zipInputStream ->
                var entry: ZipEntry? = zipInputStream.nextEntry

                while (entry != null) {
                    entryCount++
                    if (entries.size < 100) { // Limit preview to first 100 entries
                        entries.add(entry.name)
                    }
                    zipInputStream.closeEntry()
                    entry = zipInputStream.nextEntry
                }
            }

            return ArchiveInfo(
                fileName = archiveFile.name,
                fileSize = archiveFile.length(),
                entryCount = entryCount,
                previewEntries = entries
            )

        } catch (e: Exception) {
            return null
        }
    }

    data class ArchiveInfo(
        val fileName: String,
        val fileSize: Long,
        val entryCount: Int,
        val previewEntries: List<String>
    )
}
